Code/GraphMol/DistGeomHelpers/Embedder.cpp

```diff
@@ -270,19 +284,150 @@ bool _boundsFulfilled(const std::vector<int> &atoms,
+      embedParams.basinThresh));
+  unsigned int nPasses = 0;
+  field->initialize();
+  if (field->calcEnergy() > ERROR_TOL) {
+    int needMore = 1;
+    while (needMore) {
+      needMore = field->minimize(400, embedParams.optimizerForceTol);
+      ++nPasses;
+    }
+  }
+  std::vector<double> e_contribs;
+  double local_e = field->calcEnergy(&e_contribs);
+
+#ifdef DEBUG_EMBEDDING
+  std::cerr << " Energy : " << local_e / positions.size() << " "
+            << *(std::max_element(e_contribs.begin(), e_contribs.end()))
+            << std::endl;
+#endif
 
+  // check that neither the energy nor any of the contributions to it are
+  // too high (this is part of github #971)
+  if (local_e / positions->size() >= MAX_MINIMIZED_E_PER_ATOM ||
+      (e_contribs.size() &&
+       *(std::max_element(e_contribs.begin(), e_contribs.end())) >
+           MAX_MINIMIZED_E_CONTRIB)) {
+#ifdef DEBUG_EMBEDDING
+    std::cerr << " Energy fail: " << local_e / positions->size() << " "
+              << *(std::max_element(e_contribs.begin(), e_contribs.end()))
+              << std::endl;
+#endif
+    gotCoords = false;
+  }
+  return gotCoords;
+}
+
+bool checkTetrahedralCenters(const RDGeom::PointPtrVect *positions,
+                             const detail::EmbedArgs &eargs,
+                             const EmbedParameters &embedParams) {
+  RDUNUSED_PARAM(embedParams);
+  // for each of the atoms in the "tetrahedralCarbons" list, make sure
+  // that there is a minimum volume around them and that they are inside
+  // that volume. (this is part of github #971)
+  for (const auto tetSet : *eargs.tetrahedralCarbons) {
+    // it could happen that the centroid is outside the volume defined
+    // by the other
+    // four points. That is also a fail.
+    if (!_volumeTest(tetSet, *positions) ||
+        !_centerInVolume(tetSet, *positions, TETRAHEDRAL_CENTERINVOLUME_TOL)) {
+#ifdef DEBUG_EMBEDDING
+      std::cerr << " fail2! (" << tetSet->d_idx0 << ") iter: " << iter
+                << " vol: " << _volumeTest(tetSet, *positions, true)
+                << " center: "
+                << _centerInVolume(tetSet, *positions,
+                                   TETRAHEDRAL_CENTERINVOLUME_TOL, true)
+                << std::endl;
+#endif
+      return false;
+    }
+  }
+  return true;
+}
+bool checkChiralCenters(const RDGeom::PointPtrVect *positions,
+                        const detail::EmbedArgs &eargs,
+                        const EmbedParameters &embedParams) {
+  RDUNUSED_PARAM(embedParams);
+  // check the chiral volume:
+  for (const auto chiralSet : *eargs.chiralCenters) {
+    double vol = DistGeom::ChiralViolationContrib::calcChiralVolume(
+        chiralSet->d_idx1, chiralSet->d_idx2, chiralSet->d_idx3,
+        chiralSet->d_idx4, *positions);
+    double lb = chiralSet->getLowerVolumeBound();
+    double ub = chiralSet->getUpperVolumeBound();
+    if ((lb > 0 && vol < lb && (lb - vol) / lb > .2) ||
```

Comment:
Reason to use .2 here?  What is the heuristic?

Reply:
It's extremely empirical. I just looked at a bunch of structures and arrived at something that seemed to work.